### PR TITLE
CONCF-655 CuratedContent as a regular page

### PR DESCRIPTION
**Pull request by rogatty**
https://wikia-inc.atlassian.net/browse/CONCF-655

- Add new routes:
  - /main/section/:sectionName
  - /main/category/:categoryName
- Decouple main page and article - work in progress, the next step is CONCF-776
- Adjust Mercury platform

### CODE REVIEW

**Comment by tomasznapieralski** [comment]
https://github.com/Wikia/app/pull/7673

File: front/scripts/main/app.ts [hunk 1]
```diff
@@ -123,9 +121,17 @@ App.initializer({
 
 		function getPageType () {
```
Comment (tomasznapieralski) [inline front/scripts/main/app.ts]: Do we need return type?

File: front/locales/en/main.json [hunk 2]
```diff
@@ -75,6 +75,8 @@
 		"edit-publish-error-noedit-anon": "Anonymous users can't edit pages.",
 		"edit-publish-error-protectedpage": "You don't have permission to edit this page.",
 		"user-menu-profile": "Profile",
-		"user-menu-log-out": "Log out"
+		"user-menu-log-out": "Log out",
+		"category-not-exist": "Category doesn't exist",
+		"section-not-exist": "Section doesn't exist"
```
Comment (bognix) [inline front/locales/en/main.json]: @tomasznapieralski: are those translations uploaded to crowdin?

File: front/scripts/main/app.ts (hunk 1 above)
Comment (bognix) [inline front/scripts/main/app.ts]: yes

File: front/locales/en/main.json (hunk 2 above)
Comment (tomasznapieralski) [inline front/locales/en/main.json]: No they are not, i wasn't sure that text is proper?

File: front/scripts/main/app.ts (hunk 1 above)
Comment (bognix) [inline front/scripts/main/app.ts]: please check if isMainPage is available here. It would simplify logic a lot.

File: front/scripts/main/app.ts [hunk 3]
```diff
@@ -123,9 +121,17 @@ App.initializer({
...
 			var mainPageTitle = Mercury.wiki.mainPageTitle,
-				isMainPage = window.location.pathname.split('/').indexOf(mainPageTitle);
+				pathnameChunks = window.location.pathname.split('/');
+
+			if (
+				pathnameChunks.indexOf(mainPageTitle) !== -1 ||
```
Comment (bognix) [inline front/scripts/main/app.ts]: It won't work if mainPage has '/' in its name. Possibly an edge case but at least we should mention it in comment.

File: front/scripts/main/components/CuratedContentComponent.ts [hunk 4]
```diff
@@ -8,95 +8,19 @@ App.CuratedContentComponent = Em.Component.extend(App.LoadingSpinnerMixin, App.T
...
-	willDestroyElement: function (): void {
-		this.sectionsStack.clear();
-	},
+	spinnerDelay: 50,
 
 	actions: {
 		clickItem: function (item: CuratedContentItem): void {
-			if (item.type === 'section' || item.type === 'category') {
-				this.loadSection(item);
+			if (item.type) {
```
Comment (bognix) [inline front/scripts/main/components/CuratedContentComponent.ts]: move item.type to variable. We are accessing it 4 times.

File: front/scripts/main/components/MainPageComponent.ts [hunk 5]
```diff
@@ -24,17 +26,29 @@ App.MainPageComponent = Em.Component.extend({
 		}
 	}),
 
+	/**
+	 * @desc Component is reused so we have to observe on curatedContent to detect transitions between routes
+	 */
+	curatedContentObserver: Em.observer('curatedContent', function (): void {
+		// TODO: This should be refactored, ads should be initialized only once
```
Comment (bognix) [inline front/scripts/main/components/MainPageComponent.ts]: create a ticket to resolve TODO

File: front/scripts/main/models/ArticleModel.ts [hunk 6]
```diff
@@ -69,12 +69,13 @@ App.ArticleModel.reopenClass({
...
-		if (M.prop('firstPage')) {
-			this.setArticle(model);
-			return model;
-		}
-
 		return new Em.RSVP.Promise((resolve: Function, reject: Function): void => {
+			if (M.prop('firstPage')) {
+				this.setArticle(model);
+				resolve(model);
+				return;
```
Comment (tomasznapieralski) [inline front/scripts/main/models/ArticleModel.ts]: do we need return after resolve?

File: front/scripts/main/components/MainPageComponent.ts [hunk 7]
```diff
@@ -24,17 +26,29 @@ App.MainPageComponent = Em.Component.extend({
...
+
 	actions: {
 		openLightbox: function (lightboxType: string, lightboxData: any): void {
 			this.sendAction('openLightbox', lightboxType, lightboxData);
-		}
-	},
+		},
 
-	didInsertElement: function(): void {
-		M.track({
```
Comment (bognix) [inline front/scripts/main/components/MainPageComponent.ts]: I believe we should add somewhere tracking for this impression if we removed it.

File: front/scripts/main/models/CuratedContentModel.ts [hunk 8]
```diff
@@ -16,7 +15,10 @@ interface CuratedContentItem {
 	ns?: number;
 }
 
-App.CuratedContentModel = Em.Object.extend({
+// TODO this probably shouldn't be empty
```
Comment (bognix) [inline front/scripts/main/models/CuratedContentModel.ts]: create a ticket to resolve TODO

File: front/scripts/main/models/MainPageModel.ts [hunk 9]
```diff
@@ -1,14 +1,29 @@
 /// <reference path="../app.ts" />
 'use strict';
 
-App.MainPageModel = Em.Object.extend({
-	featuredContent: null,
-	curatedContent: null,
-	trendingArticles: null,
-	trendingVideos: null,
+// TODO We should decouple main page data from article but this needs a separate API path
+App.MainPageModel = App.ArticleModel.extend();
```
Comment (bognix) [inline front/scripts/main/models/MainPageModel.ts]: create a ticket to resolve TODO

File: front/scripts/main/models/MainPageModel.ts [hunk 10]
```diff
@@ -1,14 +1,29 @@
...
-		// Only curated content needs the special treatment
-		this.set('curatedContent', App.CuratedContentModel.create(this.get('curatedContent')));
+App.MainPageModel.reopenClass({
+	find: function (): Em.RSVP.Promise {
+		return new Em.RSVP.Promise((resolve: Function, reject: Function): void => {
+			var articleModel = App.ArticleModel.find({
+				basePath: Em.get(Mercury, 'wiki.basePath'),
+				title: Em.get(Mercury, 'wiki.mainPageTitle'),
+				// TODO extract this
+				wiki: Em.get(Mercury, 'wiki.dbName') || window.location.href.match(/^https?:\/\/(.*?)\./)[1]
```
Comment (bognix) [inline front/scripts/main/models/MainPageModel.ts]: create a ticket to resolve TODO

File: front/scripts/main/models/CuratedContentModel.ts [hunk 11]
```diff
@@ -29,14 +31,7 @@ App.CuratedContentModel = Em.Object.extend({
...
 				success: (data: any): void => {
-					var sanitizedData: CuratedContentItem[] = [];
-
-					if (data.items) {
-						sanitizedData = data.items.map((item: any): CuratedContentItem => {
-							return this.sanitizeItem(item);
-						});
-					}
-					resolve(sanitizedData);
+					resolve(App.CuratedContentModel.sanitizeItems(data.items || []));
```
Comment (tomasznapieralski) [inline front/scripts/main/models/CuratedContentModel.ts]: accessing section and category items will throw exception if section is empty,
`resolve(App.CuratedContentModel.sanitizeItems(data.items));` will be enough

File: front/scripts/main/routes/ArticleRoute.ts [hunk 12]
```diff
@@ -58,6 +58,24 @@ App.ArticleRoute = Em.Route.extend(App.RouteActionsMixin, {
...
+			}
+			Em.Logger.warn('Route error', error.stack || error);
+			return true;
+		},
+
+		// TODO: This currently will scroll to the top even when the app has encountered
+		// an error. Optimally, it would remain in the same place.
+		didTransition: function (): boolean {
+			// TODO: This is here only because we need to load ads resources from MW, should be done automatically
+			this.send('setupAds', this.get('controller.model.adsContext'));
```
Comment (bognix) [inline front/scripts/main/routes/ArticleRoute.ts]: create ticket to resolve TODO

File: front/scripts/main/routes/MainPageCategoryRoute.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+				mainPageData: {
+					curatedContent: model
+				}
+			}
+		})
+	},
+
+	actions: {
+		error: function (error: any, transition: EmberStates.Transition): void {
+			this.controllerFor('application').addAlert('warning', i18n.t('app.category-not-exist'));
```
Comment (bognix) [inline front/scripts/main/routes/MainPageCategoryRoute.ts]: this error should only be displayed if category doesn't exist - API returns 404.
Right now this error is triggered if any error happens. For example: TypeError. In this case user sees category content and then gets redirected to main page - really strange UX

File: front/scripts/main/routes/ApplicationRoute.ts [hunk 14]
```diff
@@ -40,7 +40,19 @@ App.ApplicationRoute = Em.Route.extend(Em.TargetActionSupport, App.TrackClickMix
 		},
 
 		error: function () {
```
Comment (tomasznapieralski) [inline front/scripts/main/routes/ApplicationRoute.ts]: return type

File: server/facets/showMainPageCategory.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+function showCategory (request: Hapi.Request, reply: Hapi.Response): void {
+	var wikiDomain: string = Utils.getCachedWikiDomainName(localSettings, request.headers.host),
+		params: MainPageRequestParams = {
+			wikiDomain: wikiDomain
+		},
+		mainPage: MainPage.MainPageRequestHelper,
+		allowCache = true;
+
+	if (request.params.categoryName) {
+		params.categoryName = request.params.categoryName.substr(request.params.categoryName.indexOf(':') + 1);
```
Comment (bognix) [inline server/facets/showMainPageCategory.ts]: This functionality is used twice. Consider moving to some helper method.

File: front/scripts/main/routes/MainPageSectionRoute.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+			model: {
+				mainPageData: {
+					curatedContent: model
+				}
+			}
+		})
+	},
+
+	actions: {
+		error: function (error: any, transition: EmberStates.Transition): void {
```
Comment (bognix) [inline front/scripts/main/routes/MainPageSectionRoute.ts]: this error should only be displayed if section doesn't exist (or contains no members) - API returns empty array.
Right now this error is triggered if any error happens. For example: TypeError. In this case user sees category content and then gets redirected to main page - really strange UX.

Best idea would be to unify API for category and section so it always returns 404 if there are no members.

File: server/facets/operations/processCuratedContentData.ts [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+		response: Hapi.Response;
+
+	result.article = result.pageData.articleData;
+	result.gridData = result.pageData.gridData;
+	delete result.pageData.gridData;
+
+	if (!result.wiki.dbName) {
+		//if we have nothing to show, redirect to our fallback wiki
+		reply.redirect(localSettings.redirectUrlOnNoData);
+	} else if ((error && error.code === 404) || (!result.gridData.items || result.gridData.items.length < 1)) {
```
Comment (bognix) [inline server/facets/operations/processCuratedContentData.ts]: this if should be simplified if we change MW API so it always returns 404 for empty sections or empty categories.
Right now it is empty array for section and 404 for category

File: front/locales/en/main.json (hunk 2 above)
Comment (Warkot) [inline front/locales/en/main.json]: I would change the the keys to include `error`, like `page-not-found-error` or `edit-publish-error-blocked`. And `curated-content` to provide more context. It could go something like this: `curated-content-error-category-not-found`. What do you think?

File: server/lib/MainPage.ts [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+				var gridDataPromise: Promise.Inspection<Promise<any>> = results[0],
+					articlePromise: Promise.Inspection<Promise<any>> = results [1],
+					wikiPromise: Promise.Inspection<Promise<any>> = results[2],
+					gridData: any,
+					articleData: any,
+					pageData: any = {},
+					wikiVariables: any = {};
+
+				// if promise is fulfilled - use resolved value, if it's not - use rejection reason
+				gridData = gridDataPromise.isFulfilled() ?
```
Comment (bognix) [inline server/lib/MainPage.ts]: rename to curatedContentData

File: server/views/_layouts/ember-main.hbs [hunk 19]
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html class="no-js" lang="{{wiki.language.user}}" dir="{{wiki.language.userDir}}">
+<html class="" lang="{{wiki.language.user}}" dir="{{wiki.language.userDir}}">
```
Comment (bognix) [inline server/views/_layouts/ember-main.hbs]: remove this empty class=""

File: front/locales/en/main.json (hunk 2 above)
Comment (vforge) [inline front/locales/en/main.json]: :+1:

File: front/scripts/main/routes/MainPageSectionRoute.ts [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+		mainPageController.setProperties({
+			isRoot: false,
+			title: sectionName,
+			adsContext: Em.get(Mercury, 'article.adsContext'),
+			ns: Em.get(Mercury, 'article.details.ns')
+		});
+	},
+
+	renderTemplate: function (controller: any, model: CuratedContentItem[]) {
```
Comment (tomasznapieralski) [inline front/scripts/main/routes/MainPageSectionRoute.ts]: return type

File: front/scripts/main/app.ts (hunk 3 above)
Comment (Warkot) [inline front/scripts/main/app.ts]: At lest it will be working consistently with what we have already. But, yeah, it would be nice to fix it using something like `isMainPage`.

File: front/scripts/main/routes/MainPageIndexRoute.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+'use strict';
+
+App.MainPageIndexRoute = Em.Route.extend({
+	model: function (): Em.RSVP.Promise {
+		return App.MainPageModel.find();
+	},
+
+	afterModel: function (model: typeof App.MainPageModel): void {
+		document.title = M.String.normalize(
+			Em.get(Mercury, 'wiki.mainPageTitle') + ' - ' + Em.getWithDefault(Mercury, 'wiki.siteName', 'Wikia')
```
Comment (vforge) [inline front/scripts/main/routes/MainPageIndexRoute.ts]: `Em.getWithDefault(Mercury, 'wiki.siteName', 'Wikia')` is not included within `M.String.normalize` anywhere else. Intended? Please fix accordingly. If intended, we also need to fix setting `document.title` anywhere else in Mercury

File: front/scripts/main/routes/MainPageRoute.ts [hunk 22]
```diff
@@ -3,20 +3,37 @@
...
+			return true;
+		},
 
-	// TODO: Main page should has its own template, we shouldn't use article template
-	renderTemplate: function (controller: any, model: typeof App.ArticleModel): void {
-		this.render('article', {
-			view: 'article',
-			model: model
-		});
+		// TODO: This currently will scroll to the top even when the app has encountered an error.
```
Comment (vforge) [inline front/scripts/main/routes/MainPageRoute.ts]: TODO is to vague - please consider adding ticket to resolve it and/or word it better.

File: front/scripts/main/components/CuratedContentComponent.ts (hunk 4 above)
Comment (Warkot) [inline front/scripts/main/components/CuratedContentComponent.ts]: Is accessing object's property that costly? Is there any complicated function running behind it?

File: front/templates/main/article.hbs [hunk 23]
```diff
@@ -1,82 +1,72 @@
...
+	{{#if model.categories}}
+		{{#collapsible-menu
+			showMenuIcon=false
+			additionalClasses='article-categories-list'
+			tLabel='app.article-categories-list-label'
+			ordered=false
+			observe=model.cleanTitle
+			trackingEvent='category'
+		}}
+			{{#each category in model.categories itemController='categoryListItem'}}
```
Comment (vforge) [inline front/templates/main/article.hbs]: Please use new syntax `#each X as |Y|` instead of `#each Y in X`

Comment (rogatty) [inline front/templates/main/article.hbs]: This is already done in https://github.com/Wikia/mercury/pull/919. Let's not touch things that are not related to the ticket (only indentation changed here but git handles it poorly).

File: front/scripts/main/app.ts [hunk 24]
```diff
@@ -40,9 +40,9 @@ App.initializer({
 	name: 'preload',
 	initialize: (container: any, application: any) => {
 		var debug: boolean = M.prop('environment') === 'dev',
-			//prevents fail if transitions are empty
+		//prevents fail if transitions are empty
 			loadedTranslations = M.prop('translations') || {},
-			//loaded language name is the first key of the Mercury.state.translations object
+		//loaded language name is the first key of the Mercury.state.translations object
```
Comment (rogatty) [inline front/scripts/main/app.ts]: It was better before.

File: front/scripts/main/routes/MainPageRoute.ts (hunk 22 above)
Comment (tomasznapieralski) [inline front/scripts/main/routes/MainPageRoute.ts]: Added

File: front/scripts/main/app.ts (hunk 1 above)
Comment (rogatty) [inline front/scripts/main/app.ts]: It's not available through `ArticleModel` because initializers run before the rest of Ember starts. There might be a way to get it from `M.prop` but this feels hacky and will stop working after https://github.com/Wikia/mercury/pull/999 is merged and enabled (article data will be loaded asynchronously).

Comment (tomasznapieralski) [inline front/scripts/main/app.ts]: I will add return type only for now

File: front/scripts/main/components/CuratedContentComponent.ts (hunk 4 above)
Comment (rogatty) [inline front/scripts/main/components/CuratedContentComponent.ts]: This is microoptimisation but might look better (shorter lines).

File: front/scripts/main/app.ts (hunk 24 above)
Comment (Warkot) [inline front/scripts/main/app.ts]: +1

**Comment by rogatty** [comment]
@gabrys @nandy-andy can the AdEng team take a look at this? Just search for "ads" ;)

File: front/scripts/main/components/MainPageComponent.ts (hunk 7 above)
Comment (rogatty) [inline front/scripts/main/components/MainPageComponent.ts]: We've agreed to keep it removed as we're tracking page views for main pages already.

File: server/facets/showMainPageCategory.ts (hunk 15 above)
Comment (rogatty) [inline server/facets/showMainPageCategory.ts]: You mean `categoryName = item.categoryName.substr(item.categoryName.indexOf(':') + 1);` in `MainPageRoute.ts`?
If so then we can't - there is no common code base between back and front ends.

Altough we could make sure that we pass already sanitized category name from the front end, couldn't we?

File: front/templates/main/article.hbs (hunk 23 above)
Comment (vforge) [inline front/templates/main/article.hbs]: :+1:

File: front/scripts/main/routes/MainPageIndexRoute.ts (hunk 21 above)
Comment (vforge) [inline front/scripts/main/routes/MainPageIndexRoute.ts]: ping @Warkot

File: front/scripts/main/routes/MainPageSectionRoute.ts (hunk 16 above)
Comment (rogatty) [inline front/scripts/main/routes/MainPageSectionRoute.ts]: Fixed in https://github.com/Wikia/mercury/commit/d6e6f9b03b885c6b04acf4dff1191d07478217e1.

File: front/scripts/main/controllers/MainPageController.ts [hunk 25]
```diff
@@ -0,0 +1,16 @@
+/// <reference path="../app.ts" />
+/// <reference path="../../baseline/mercury.d.ts" />
+'use strict';
+
+App.MainPageController = Em.Controller.extend({
+	needs: ['application'],
+
+	noAds: Em.computed.alias('controllers.application.noAds'),
```
Comment (nandy-andy) [inline front/scripts/main/controllers/MainPageController.ts]: What does it do here? Shouldn't it also be refactored (maybe even within ADEN-2189 mentioned [here](https://github.com/Wikia/mercury/pull/993/files#diff-ca00dc38f40fdedb6e32ced5881c292aR33))? I imagine there is a higher level on which we decide to load or not ads - `ApplicationController` or something...

Comment (rogatty) [inline front/scripts/main/controllers/MainPageController.ts]: It's here because `noAds` query param is a property of `ApplicationController` and we need to have it in `MainPageComponent` to setup ads. We can't do it in the route or controller because `injectAds` and `injectMainPageAds` need a rendered view to inject ads dynamically.

**Comment by rogatty** [comment]
LGTM

**Comment by vforge** [comment]
LGTM

**Comment by rogatty** [comment]
@kenkouot @lizlux according to:
>Platform work can be implemented by other teams with architectural/code review by Mercury team

I'm asking you to take a look at this PR as we're touching some parts of the core.

**Comment by bognix** [comment]
LGTM but please create separate issue for finding better solution. Let's put it in the next sprint.

**Comment by vforge** [comment]
LGTM

**Comment by tomasznapieralski** [comment]
LGTM

File: test/specs/front/scripts/main/models/ArticleModel.js [hunk 26]
```diff
@@ -82,25 +74,27 @@ test('setArticle with parametrized data', function () {
...
+			article: 'article'
+		};
 
 	ok(M.prop('firstPage'), 'firstPage==true before test, as expected');
 	Ember.run(function () {
 		model = App.ArticleModel.find(params);
 	});
-	verifyArticle(model, this.example, this.wikiExample);
+	model.then(function (resolvedModel) {
+		verifyArticle(resolvedModel, self.example);
```
Comment (Warkot) [inline test/specs/front/scripts/main/models/ArticleModel.js]: Doesn't `function () =>…` handle the `self = this` thing?

Comment (rogatty) [inline test/specs/front/scripts/main/models/ArticleModel.js]: It does but we don't use TS in tests, it's a JS file.

**Comment by Warkot** [comment]
LGTM

File: test/specs/front/scripts/main/models/ArticleModel.js (hunk 26 above)
Comment (vforge) [inline test/specs/front/scripts/main/models/ArticleModel.js]: one more reason for ES6 :cat2:

Comment (Warkot) [inline test/specs/front/scripts/main/models/ArticleModel.js]: M'kay. :dog2:

**Comment by lizlux** [comment]
LGTM
